File: array/group-anagrams.js
/**
 * https://leetcode.com/problems/group-anagrams
 * 
 * @param {string[]} strs
 * @return {string[][]}
 */
const groupAnagrams = (strs) => {
    const letterMap = new Map();

    for (str of strs) {
        const sortedStr = str.split('').sort().join('')

        if (!letterMap.has(sortedStr)) {
            letterMap.set(sortedStr, [str]);
        } else {
            letterMap.get(sortedStr).push(str);
        }
    };

    return Array.from(letterMap.values());
};

//Builds on the valid anagram question, we use same trick to find an anagram,
//then we have to use an ordered collection to group values together

File: python/array/group-anagrams.py
class Solution:
    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:
        groups = {}

        if not len(strs):
            return [strs[0]]

        for i in strs:
            sorted_str = ''.join(sorted(i))
            if sorted_str in groups:
                groups[sorted_str].append(i)
            else:
                groups[sorted_str] = [i]

        return groups.values()